src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -469,8 +469,14 @@ private void Register(Interop.Sys.SocketEvents events)
                 Interop.Error errorCode;
                 if (!_asyncEngineToken.TryRegister(_socket, _registeredEvents, events, out errorCode))
                 {
-                    // TODO (#7850): throw an appropriate exception
-                    throw new Exception(string.Format("SocketAsyncContext.Register: {0}", errorCode));
+                    if (errorCode == Interop.Error.ENOMEM || errorCode == Interop.Error.ENOSPC)
+                    {
+                        throw new OutOfMemoryException();
+                    }
+                    else
+                    {
+                        throw new InternalException();                        
```

Comment:
This is a general issue with InternalException, which we should fix everywhere.  I've filed #8517 to track this.